[C#] Add inheritance exercise (#1424)
exercism/v3: languages/csharp/exercises/concept/inheritance/.docs/instructions.md

- They have a number of hit points.
- They can attack, which reduces the number of hit points of the other character.
- If the number of hit points is less than or equal to zero, they are stunned and cannot do any more damage.

There are also some differences between the two character types.

|                    | Warriors          | Wizards             |
| ------------------ | ----------------- | ------------------- |
| Initial hit points | 30                | 20                  |
| Default attack     | 6                 | 3                   |
| Special attack     | 10 (potion drunk) | 12 (spell prepared) |

Drinking a potion (warrior) or preparing a spell (wizard) only makes the next attack a special attack. Subsequent attacks will be default attacks, unless a new potion is drunk or a new spell is prepared.

You have six tasks that work with warriors and wizard characters.

### 1. Display the hit points

Override the `ToString()` method on the `Character` class to display the number of hit points of a character:

```csharp
var warrior = new Warrior();
warrior.ToString();
// => "HP: 30"
```

### 2. Allow attacking a character using a basic attack

> Implement the `Warrior.Damage()` and `Wizard.Damage()` methods to return the damage done by their character type's basic attack. Next, implement the `Character.Attack()` method that takes the character to attack as its paramete and decrease the attacked character's hit points with the attacking character's damage.

Reviewer: ```suggestion
Implement the `Warrior.Damage()` and `Wizard.Damage()` methods to return the damage done by their character type's basic attack. Next, implement the `Character.Attack()` method that takes the character to attack as its parameter and decrease the attacked character's hit points with the attacking character's damage.
```

> Implement the `Warrior.Damage()` and `Wizard.Damage()` methods to return the damage done by their character type's basic attack. Next, implement the `Character.Attack()` method that takes the character to attack as its paramete and decrease the attacked character's hit points with the attacking character's damage.

Reviewer: Standardise on "Default attack" or "Basic attack"

Author: I've standardized on Basic Attack.

```csharp
var warrior = new Warrior();
var wizard = new Wizard();

wizard.PrepareSpell();
wizard.Attack(warrior);
warrior.ToString();
// => "HP: 18"
```

### 4. Allow warriors to use their special attack

Implement the `Warrior.DrinkPotion()` method to allow a warrior to drink their potion, making their next attack a special attack. Update the `Warrior.Damage()` method to return the special attack damage if a potion has been drunk.

```csharp
var warrior = new Warrior();
var wizard = new Wizard();

warrior.DrinkPotion();
warrior.Attack(wizard);
wizard.ToString();
// => "HP: 10"
```

### 5. Check for stunned characters

Implement the `Character.Stunned()` method to return `true` if the character's hit points are less than or equal to zero:

```csharp
var warrior = new Warrior();
var wizard = new Wizard();
warrior.Attack(wizard);
warrior.Attack(wizard);
warrior.Attack(wizard);
warrior.Attack(wizard);
wizard.Stunned();
// => true
```

### 6. Stunned characters cannot do damage

> Update the `Character.Attack()` method to not do any damage to another character if this character is stunned:

Reviewer: ```suggestion
Update the `Character.Attack()` so that a Stunned character cannot cause damage.
```

("this character" was ambigous to me)

In this exercise you're playing a role-playing game named "Wizard and Warriors," which allows you to play as either a Wizard or a Warrior.

There are different rules for Warriors and Wizards to determine how much damage points they deal.

For a Warrior, these are the rules:

> - Deal 6 points of damage if the attacked character is not vulnerable

Reviewer: ```suggestion
- Deal 6 points of damage if the character they are attacking is not vulnerable
```

> - Deal 10 points of damage if the attacked character is vulnerable

Reviewer: ```suggestion
- Deal 10 points of damage if the character they are attacking is vulnerable
```

For a Wizard, these are the rules:

- Deal 3 points of damage if the Wizard did not prepare a spell in advance

> - Deal 12 points of damage if the Wizard did prepare a spell in advanced

Reviewer: ```suggestion
- Deal 12 points of damage if the Wizard prepared a spell in advanced
```

> - Deal 3 points of damage if the Wizard did not prepare a spell in advance
> - Deal 12 points of damage if the Wizard did prepare a spell in advanced

Reviewer: ```suggestion
- Deal 12 points of damage if the Wizard prepared a spell in advanced
- Deal 3 points of damage if the Wizard did not prepare a spell in advance
```
